### PR TITLE
[AIRFLOW-4135] Add Google Cloud Build operator and hook

### PR DESCRIPTION
**Pull request by mik-laj**
Make sure you have checked _all_ steps below.

### Jira

  - https://issues.apache.org/jira/browse/AIRFLOW-4135

### Description

Hello,

I want to create a new operator and hook for Google Cloud Build.

Operator name | API URL /description
-- | --
CloudBuildCreateBuildOperator | [Python API](https://developers.google.com/resources/api-libraries/documentation/cloudbuild/v1/python/latest/cloudbuild_v1.projects.builds.html#create)

Thanks,

### Tests

Unit and system tests have been added.

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [ ] Passes `flake8`

### CODE REVIEW

**Comment by OmerJog** [comment]
Missing `integration.rst` addition? :)

**Comment by mik-laj** [comment]
@OmerJog good catch :-D .I updated a PR.  

Thank you for your work in the community. I'm glad you want to work with me

File: airflow/contrib/hooks/gcp_cloud_build_hook.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        :return: None
+        """
+        service = self.get_conn()
+
+        # Create build
+        response = (
+            service.projects()
+            .builds()
+            .create(projectId=project_id, body=body)
+            .execute(num_retries=NUM_RETRIES)
```
Comment (milton0825) [inline airflow/contrib/hooks/gcp_cloud_build_hook.py]: Should we allow users to override the default `NUM_RETRIES`?

File: airflow/contrib/hooks/gcp_cloud_build_hook.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        asynchronous call.
+
+        :param operation_name: The name of the operation.
+        :type operation_name: str
+        :return: The response returned by the operation.
+        :rtype: dict
+        :exception: AirflowException in case error is returned.
+        """
+        service = self.get_conn()
+        while True:
```
Comment (milton0825) [inline airflow/contrib/hooks/gcp_cloud_build_hook.py:119]: Should we put a timeout here?

File: tests/contrib/hooks/test_gcp_cloud_build_hook.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        execute_mock = mock.Mock(**{"side_effect": [
+            TEST_WAITING_OPERATION,
+            TEST_ERROR_OPERATION,
+        ]})
+        service_mock.operations.return_value.get.return_value.execute = execute_mock
+        with six.assertRaisesRegex(self, AirflowException, "error"):
+            self.hook.create_build(body={})
+
+
+class TestGcpComputeHookNoDefaultProjectId(unittest.TestCase):
```
Comment (sdevani) [inline tests/contrib/hooks/test_gcp_cloud_build_hook.py]: No test cases for this class and the one below?

File: tests/contrib/hooks/test_gcp_cloud_build_hook.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    def setUp(self):
+        with mock.patch(
+            "airflow.contrib.hooks.gcp_api_base_hook.GoogleCloudBaseHook.__init__",
+            new=mock_base_gcp_hook_default_project_id,
+        ):
+            self.hook = CloudBuildHook(gcp_conn_id="test")
+
+    @mock.patch('airflow.contrib.hooks.gcp_cloud_build_hook.CloudBuildHook.get_conn')
+    def test_minimal_green_path(self, get_conn_mock):
```
Comment (sdevani) [inline tests/contrib/hooks/test_gcp_cloud_build_hook.py]: nit: test_build_immediately_complete would be more descriptive of what this is testing.

File: tests/contrib/operators/test_gcp_cloud_build_operator.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        BuildProcessor(body=body).process_body()
+        self.assertEquals(body, expected_body)
+
+
+class GcpCloudBuildCreateBuildOperatorTestCase(TestCase):
+    @mock.patch(
+        "airflow.contrib.operators.gcp_cloud_build_operator.CloudBuildHook",
+        **{"return_value.create_build.return_value": TEST_CREATE_BODY}
+    )
+    def test_minimal_green_path(self, hock_mock):
```
Comment (sdevani) [inline tests/contrib/operators/test_gcp_cloud_build_operator.py]: nit: did you mean hook_mock?

File: tests/contrib/operators/test_gcp_cloud_build_operator.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+            # [END howto_operator_gcp_cloud_build_source_gcs_dict]
+            "steps": [
+                {
+                    "name": "gcr.io/cloud-builders/docker",
+                    "args": ["build", "-t", "gcr.io/$PROJECT_ID/docker-image", "."],
+                }
+            ],
+            "images": ["gcr.io/$PROJECT_ID/docker-image"],
+        }
+        hock_mock.create_build(body=expected_result, project_id=TEST_PROJECT_ID)
```
Comment (sdevani) [inline tests/contrib/operators/test_gcp_cloud_build_operator.py]: There is no assertion in this test

Comment (sdevani) [reply tests/contrib/operators/test_gcp_cloud_build_operator.py]: Same with test below

**Review by sdevani — CHANGES_REQUESTED** [review]

File: airflow/contrib/hooks/gcp_cloud_build_hook.py (hunk 2 above)
Comment (mik-laj) [reply airflow/contrib/hooks/gcp_cloud_build_hook.py:119]: There is a timeout for the duration of the entire operator. I think that there is no need to add another timeout. WDYT?

Comment (mik-laj) [reply airflow/contrib/hooks/gcp_cloud_build_hook.py:119]: https://github.com/apache/airflow/blob/master/airflow/contrib/hooks/gcp_compute_hook.py#L290-L320
I found a similar code in project. I prefer to keep the consequences in the project.

File: tests/contrib/hooks/test_gcp_cloud_build_hook.py (hunk 4 above)
Comment (mik-laj) [reply tests/contrib/hooks/test_gcp_cloud_build_hook.py]: Thanks. I updated the PR.

File: tests/contrib/operators/test_gcp_cloud_build_operator.py (hunk 5 above)
Comment (mik-laj) [reply tests/contrib/operators/test_gcp_cloud_build_operator.py]: Thanks. I updated the PR.

File: tests/contrib/operators/test_gcp_cloud_build_operator.py (hunk 6 above)
Comment (mik-laj) [reply tests/contrib/operators/test_gcp_cloud_build_operator.py]: Thanks. I updated the PR.

File: tests/contrib/hooks/test_gcp_cloud_build_hook.py (hunk 3 above)
Comment (mik-laj) [reply tests/contrib/hooks/test_gcp_cloud_build_hook.py]: Thanks. I updated the PR.

File: airflow/contrib/hooks/gcp_cloud_build_hook.py (hunk 1 above)
Comment (mik-laj) [reply airflow/contrib/hooks/gcp_cloud_build_hook.py]: Oh. Of course. It's a good idea. I did not notice that it was already introduced

**Comment by mik-laj** [comment]
@potiuk @kaxil PTAL

File: airflow/contrib/hooks/gcp_cloud_build_hook.py (hunk 2 above)
Comment (potiuk) [reply airflow/contrib/hooks/gcp_cloud_build_hook.py:119]: Each request issued by execute() method in discovery api has default timeout = 60s (per retry).

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
@mik-laj - can you please rebase and get rid of the conflict ?

**Comment by mik-laj** [comment]
@potiuk  I updated the PR.

**Pull request merged**